zebra-chain/src/value_balance.rs

```diff
@@ -0,0 +1,248 @@
+        ValueBalance {
+            sprout: sprout_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given sapling amount.
+    pub fn from_sapling_amount(sapling_amount: Amount<C>) -> Self {
+        ValueBalance {
+            sapling: sapling_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given orchard amount.
+    pub fn from_orchard_amount(orchard_amount: Amount<C>) -> Self {
+        ValueBalance {
+            orchard: orchard_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Insert a transparent value balance into a given `ValueBalance`
+    /// leaving the other values untouched.
+    pub fn transparent_value_balance(
+        &mut self,
+        transparent_value_balance: ValueBalance<C>,
+    ) -> &Self {
+        self.transparent = transparent_value_balance.transparent;
+        self
+    }
+
+    /// Insert a sprout value balance into a given `ValueBalance`
+    /// leaving the other values untouched.
+    pub fn sprout_value_balance(&mut self, sprout_value_balance: ValueBalance<C>) -> &Self {
+        self.sprout = sprout_value_balance.sprout;
+        self
+    }
+
+    /// Insert a sapling value balance into a given `ValueBalance`
+    /// leaving the other values untouched.
+    pub fn sapling_value_balance(&mut self, sapling_value_balance: ValueBalance<C>) -> &Self {
+        self.sapling = sapling_value_balance.sapling;
+        self
+    }
+
+    /// Insert an orchard value balance into a given `ValueBalance`
+    /// leaving the other values untouched.
+    pub fn orchard_value_balance(&mut self, orchard_value_balance: ValueBalance<C>) -> &Self {
+        self.orchard = orchard_value_balance.orchard;
+        self
+    }
+}
+
+impl<C> Default for ValueBalance<C>
+where
+    C: Constraint + Copy,
+{
+    fn default() -> Self {
+        let zero = Amount::zero();
+        Self {
+            transparent: zero,
+            sprout: zero,
+            sapling: zero,
+            orchard: zero,
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug, Clone, PartialEq)]
+//#[allow(missing_docs)]
+/// Errors that can be returned when validating `ValueBalance`s
+enum ValueBalanceError {
+    #[error("value balance contains invalid amounts")]
+    AmountError(#[from] crate::amount::Error),
+}
+
+impl<C> std::ops::Add<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+
+    fn add(self, rhs: ValueBalance<C>) -> Self::Output {
+        let value_balance = self?;
+
+        Ok(ValueBalance::<C> {
+            transparent: (value_balance.transparent + rhs.transparent)?,
+            sprout: (value_balance.sprout + rhs.sprout)?,
+            sapling: (value_balance.sapling + rhs.sapling)?,
+            orchard: (value_balance.orchard + rhs.orchard)?,
+        })
+    }
+}
```

Comment:
Minor: would it make sense to add `impl Add for ValueBalance` as well?

```suggestion
impl<C> std::ops::Add for ValueBalance<C>
where
    C: Constraint,
{
    type Output = Result<ValueBalance<C>, ValueBalanceError>;

    fn add(self, rhs: ValueBalance<C>) -> Self::Output {
        Ok(ValueBalance::<C> {
            transparent: (self.transparent + rhs.transparent)?,
            sprout: (self.sprout + rhs.sprout)?,
            sapling: (self.sapling + rhs.sapling)?,
            orchard: (self.orchard + rhs.orchard)?,
        })
    }
}

impl<C> std::ops::Add<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
where
    C: Constraint,
{
    type Output = Result<ValueBalance<C>, ValueBalanceError>;

    fn add(self, rhs: ValueBalance<C>) -> Self::Output {
        self? + rhs
    }
}
```

Reply:
Yes, we did something similar when implementing `Sum` for `Amount` so i think it makes sense to do this in here as well.

I implemented this as suggested in https://github.com/ZcashFoundation/zebra/pull/2505/commits/593eb7c011f7fa73f6d6be5193b05e5f3e083ffd

Reply:
I added this also to `Sub` in the context of https://github.com/ZcashFoundation/zebra/pull/2505/commits/4b5f1e64489ecc90d81c64411f8ffa0e6880c97f